src/PowerShellEditorServices/Session/PSReadLinePromptContext.cs

```diff
@@ -0,0 +1,197 @@
+    using System.Management.Automation;
+
+    internal class PSReadLinePromptContext : IPromptContext {
+        private const string ReadLineScript = @"
+            [System.Diagnostics.DebuggerHidden()]
+            [System.Diagnostics.DebuggerStepThrough()]
+            param()
+            return [Microsoft.PowerShell.PSConsoleReadLine]::ReadLine(
+                $Host.Runspace,
+                $ExecutionContext,
+                $args[0])";
+
+        private const string ReadLineInitScript = @"
+            [System.Diagnostics.DebuggerHidden()]
+            [System.Diagnostics.DebuggerStepThrough()]
+            param()
+            end {
+                $module = Get-Module -ListAvailable PSReadLine | Select-Object -First 1
+                if (-not $module -or $module.Version -lt ([version]'2.0.0')) {
+                    return
+                }
+
+                Import-Module -ModuleInfo $module
+                return 'Microsoft.PowerShell.PSConsoleReadLine' -as [type]
+            }";
+
+        private readonly PowerShellContext _powerShellContext;
+
+        private PromptNest _promptNest;
+
+        private InvocationEventQueue _invocationEventQueue;
+
+        private ConsoleReadLine _consoleReadLine;
+
+        private CancellationTokenSource _readLineCancellationSource;
+
+        private PSReadLineProxy _readLineProxy;
+
+        internal PSReadLinePromptContext(
+            PowerShellContext powerShellContext,
+            PromptNest promptNest,
+            InvocationEventQueue invocationEventQueue,
+            PSReadLineProxy readLineProxy)
+        {
+            _promptNest = promptNest;
+            _powerShellContext = powerShellContext;
+            _invocationEventQueue = invocationEventQueue;
+            _consoleReadLine = new ConsoleReadLine(powerShellContext);
+            _readLineProxy = readLineProxy;
+
+#if CoreCLR
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                return;
+            }
+
+            _readLineProxy.OverrideReadKey(
+                intercept => ConsoleProxy.UnixReadKey(
+                    intercept,
+                    _readLineCancellationSource.Token));
+#endif
+        }
+
+        internal static bool TryGetPSReadLineProxy(Runspace runspace, out PSReadLineProxy readLineProxy)
+        {
+            readLineProxy = null;
+            using (var pwsh = PowerShell.Create())
+            {
+                pwsh.Runspace = runspace;
+                var psReadLineType = pwsh
+                    .AddScript(ReadLineInitScript)
+                    .Invoke<Type>()
+                    .FirstOrDefault();
+
+                if (psReadLineType == null)
+                {
+                    return false;
+                }
+
+                try
+                {
+                    readLineProxy = new PSReadLineProxy(psReadLineType);
+                }
+                catch (InvalidOperationException)
+                {
```

Comment:
Yeah I was just thinking about that.  I didn't do it originally because I knew it'd be better to tighten up the import so it picks the best module to import.  That still needs to happen, but I'm realizing we'll always want the logs anyway.